### PR TITLE
added events data functions

### PR DESCRIPTION
**Pull request by ryan280**

### CODE REVIEW

File: server/data/events.ts [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    /**
+     * Get event by id
+     * @param {ObjectId} cycleId
+     * @param {ObjectId} applicationId
+     * @param {ObjectId} eventId
+     * @returns {Promise<object>} The event given the ids above
+     */
+    readById: async (cycleId: ObjectId, applicationId: ObjectId, eventId: ObjectId):Promise<object> => {
+
+        const cyclesCollection: any = await collections.cycles();
```
Comment (GraceMattern) [inline server/data/events.ts:15]: Should use getCycleById in data/cycle

File: server/data/events.ts [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        const cycle = await cyclesCollection.findOne({_id: new ObjectId(cycleId)});
```
Comment (GraceMattern) [inline server/data/events.ts]: Should use getCycleById in data/cycle

File: server/data/events.ts [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+export default {
+    /**
+     * Get event by id
+     * @param {ObjectId} cycleId
+     * @param {ObjectId} applicationId
+     * @param {ObjectId} eventId
+     * @returns {Promise<object>} The event given the ids above
+     */
+    readById: async (cycleId: ObjectId, applicationId: ObjectId, eventId: ObjectId):Promise<object> => {
+
```
Comment (GraceMattern) [inline server/data/events.ts]: error check ie checkObjectId(cycleId), checkObjectId(applicationId), etc

File: server/data/events.ts [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        const cyclesCollection: any = await collections.cycles();
+        const cycle = await cyclesCollection.findOne({_id: new ObjectId(cycleId)});
+        if(cycle === null) throw new Error('Could not find cycle');
+
+        let application = null;
+        for(let element of cycle.applications){
+            if(element._id.toString() === applicationId.toString()){
```
Comment (GraceMattern) [inline server/data/events.ts]: Could use readById from data/applications to get application by id

File: server/data/events.ts [hunk 5]
```diff
@@ -0,0 +1,224 @@
+import collections from '../config/mongoCollections';
```
Comment (GraceMattern) [inline server/data/events.ts]: I think you might need to edit config/mongoCollections and data/index as well

File: server/data/events.ts [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+    /**
+     * Get all events
+     * @param {ObjectId} cycleId
+     * @param {ObjectId} applicationId
+     * @returns {Promise<object[]>}All the events of the application
+     */
+    readAll: async (cycleId: ObjectId, applicationId: ObjectId): Promise<object[]> => {
+
+        const cyclesCollection: any = await collections.cycles();
+        const cycle = await cyclesCollection.findOne({_id: new ObjectId(cycleId)});
```
Comment (GraceMattern) [inline server/data/events.ts]: same comment as before (regard getCycleById)

File: server/data/events.ts [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+        if(cycle === null) throw new Error('Could not find cycle');
+
+        let application = null;
+        for(let element of cycle.applications){
+            if(element._id.toString() === applicationId.toString()){
```
Comment (GraceMattern) [inline server/data/events.ts]: same comment as before (regarding getApplicationById)

File: server/data/events.ts [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+        return event;
+    },
+
+    /**
+     * Get all events
+     * @param {ObjectId} cycleId
+     * @param {ObjectId} applicationId
+     * @returns {Promise<object[]>}All the events of the application
+     */
+    readAll: async (cycleId: ObjectId, applicationId: ObjectId): Promise<object[]> => {
```
Comment (GraceMattern) [inline server/data/events.ts]: must error check

File: server/data/events.ts [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+        for(let element of cycle.applications){
+            if(element._id.toString() === applicationId.toString()){
+                application = element;
+                break;
+            }
+        }
+        if(application === null) throw new Error('Could not find application');
+        
+        let events = application.events.map((element) => element.toString());
+
```
Comment (GraceMattern) [inline server/data/events.ts]: probably a good idea to check !events

File: server/data/events.ts [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+    /**
+     * Create event
+     * @param {ObjectId} cycleId 
+     * @param {string} title 
+     * @param {Date} date 
+     * @param {String} location 
+     * @returns {Promise<object>} if the event was created, throws otherwise 
+     */
+    create: async (cycleId: ObjectId, applicationId: ObjectId, title: string, date: Date, location: string): Promise<object> => {
```
Comment (GraceMattern) [inline server/data/events.ts]: must error check

File: server/data/events.ts [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+
+        const cyclesCollection: any = await collections.cycles();
+        const cycle = await cyclesCollection.findOne({_id: new ObjectId(cycleId)});
```
Comment (GraceMattern) [inline server/data/events.ts]: same comment as before (regarding getCycleById)

File: server/data/events.ts [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+    },
+
+    /**
+     * Update event
+     * @param {ObjectId} cycleId 
+     * @param {ObjectId} applicationId 
+     * @param {ObjectId} eventId
+     * @param {object} eventObject can contain: {title: string, date: Date, location: string}
+     */
+    update: async (cycleId: ObjectId, applicationId: ObjectId, eventId: ObjectId, eventObject: object): Promise<object> => {
```
Comment (GraceMattern) [inline server/data/events.ts]: error check

File: server/data/events.ts [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+
+
+        const cyclesCollection: any = await collections.cycles();
+        const cycle = await cyclesCollection.findOne({_id: new ObjectId(cycleId)});
```
Comment (GraceMattern) [inline server/data/events.ts]: getCycleById

File: server/data/events.ts [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+        if(cycle === null) throw new Error('Could not find cycle');
+
+        //find the application
+        let application = null;
+        for(let element of cycle.applications){
+            if(element._id.toString() === applicationId.toString()){
```
Comment (GraceMattern) [inline server/data/events.ts]: getApplicationById

File: server/data/events.ts [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+    },
+
+    /**
+     * Delete event
+     * @param {ObjectId} cycleId 
+     * @param {ObjectId} applicationId 
+     * @param {ObjectId} eventId 
+     * @returns true if the event was deleted, throws otherwise
+     */
+    delete: async (cycleId: ObjectId, applicationId: ObjectId, eventId: ObjectId): Promise<boolean> => {
```
Comment (GraceMattern) [inline server/data/events.ts]: same series of comments
- error check
- if you can try to reuse code that's already been written

File: server/data/events.ts (hunk 10 above)
Comment (roccopolimen) [inline server/data/events.ts]: JS doc does not match actual function. Also cycleId, applicationId, etc will be passed as strings and not as ObjectIds. They get converted to those types on this layer (data)

**Review by roccopolimen — COMMENTED** [review]
Quick comment made in passing, as I write the routes using some of this logic.

File: server/data/events.ts (hunk 10 above)
Comment (roccopolimen) [reply server/data/events.ts]: Also cycleId is not gonna get provided, they can only add events for the users current cycle. Instead the routes will be giving you the current userId and based on that you grab their current cycle and handle from there.

Comment (ryan280) [reply server/data/events.ts]: Agree with everything besides the cycleId not being passed. There is no indication in the users collection, or cycles collection for the users current cycle.

File: server/data/events.ts (hunk 9 above)
Comment (ryan280) [reply server/data/events.ts]: agreed

**Comment by roccopolimen** [comment]
Still using generic object for typing except for 1 single time you use EventObject...

**Comment by roccopolimen** [comment]
`import { cycles } from '../config/mongoCollections';`
instead of importing collections and using collection.cycles() just use cycles() directly to tighten scope of modules loaded into codebase.

Still asking for cycleId instead of userId (check slack we discussed the proper way we are handling getting the current cycle).

instead of export default {...}, better practice is to export const <function-name> = async () => {};
for each function you created. Check both the cycles and users data for examples of this.

Also do not import typings you do not end up using.

**Comment by roccopolimen** [comment]
still not using the EventObject type where it should be used, but I am impatient at this point, so might just merge, and change myself.

**Pull request merged**